ipapython/ipautil.py

```diff
@@ -282,6 +282,37 @@ def write_tmp_file(txt):
 
     return fd
 
+
+@contextmanager
+def concurrent_open(path, mode='r'):
+    """Ensure that complete file is read/written
+
+    Works only for r, rb, w, w+, wb, wb+ modes. Default is r.
+
+    In read mode behaves the same as build-in open. In write mode all data are
+    written to the temporary file first and then moved to target path.
+    This approach ensures that reading will be performed on complete file and
+    writting will result in complete or none file written at all.
+    """
+    if mode in ('w', 'w+', 'wb', 'wb+'):
+        with tempfile.NamedTemporaryFile(
+                mode=mode, prefix=path, delete=False) as temp_file:
+            yield temp_file
+            temp_file.flush()
+            os.fsync(temp_file.fileno())
+            try:
+                os.rename(temp_file.name, path)
+            except EnvironmentError:
```

Comment:
Since I've first read PEP 3151 (expecially section about Confusing OS-related exceptions [1])  I like to play it safely and always catch EnvironmentError because (like to PEP says):
In fact, it is hard to think of any situation where OSError should be caught but not IOError, or the reverse.

[1] https://www.python.org/dev/peps/pep-3151/#confusing-set-of-os-related-exceptions